src/Microsoft.AspNetCore.Routing.Abstractions/Endpoint.cs

```diff
@@ -8,15 +8,21 @@ namespace Microsoft.AspNetCore.Routing
     [DebuggerDisplay("{DisplayName,nq}")]
     public abstract class Endpoint
     {
-        protected Endpoint(EndpointMetadataCollection metadata, string displayName)
+        protected Endpoint(
+            EndpointMetadataCollection metadata,
+            string displayName,
+            Address address)
         {
-            // Both allowed to be null
+            // All are allowed to be null
             Metadata = metadata ?? EndpointMetadataCollection.Empty;
             DisplayName = displayName;
+            Address = address;
         }
 
         public string DisplayName { get; }
 
         public EndpointMetadataCollection Metadata { get; }
+
+        public Address Address { get; }
```

Comment:
I view Address here as kind of a metadata. It can be extended by users who could write a custom finder to look up apart from name/methodinfo

Reply:
Do you think it's fundamental that `Endpoint -> Address` is `1:1?` 

If I wrote a custom finder, how does that address get associated with the endpoint?

Reply:
If it is metadata should it be something that gets added to the metadata collection?

Reply:
This to me seems like something that will grow up as we put more into this.